File: java-vfs/src/main/java/emg/java/vfs/files/DirEntry.java
package emg.java.vfs.files;

import emg.java.vfs.filesystem.FileSystemException;

public abstract class DirEntry {
    protected String parentPath;
    protected String name;

    public DirEntry(String parentPath, String name) {
        this.parentPath = parentPath;
        this.name = name;
    }

    public String parentPath() {
        return parentPath;
    }

    public String name() {
        return name;
    }

    public String path() {
        var separator = "";
        if (!Directory.ROOT_PATH.equals(parentPath)) {
            separator = Directory.SEPARATOR;
        }
        return parentPath + separator + name;
    }

    public Directory asDirectory() {
        throw new FileSystemException(String.format("%s is not a Directory", name));
    }

    public File asFile() {
        throw new FileSystemException(String.format("%s is not a File", name));
    }

    public abstract String getType();

    public boolean isFile() {
        return false;
    }

    public boolean isDirectory() {
        return false;
    }
}

File: scala-vfs/src/main/scala/emg/scala/vfs/files/DirEntry.scala
package emg.scala.vfs.files

import java.nio.file.FileSystemException

abstract class DirEntry(val parentPath: String, val name: String) {
  def path: String = parentPath + Directory.SEPARATOR + name

  def asDirectory: Directory = throw new FileSystemException(s"$name is not a Directory")

  def asFile: File = throw new FileSystemException(s"$name is not a File")

  def getType: String
}
